### PR TITLE
short circuit remaining validation rules if one exceeds its per-cost limit

### PR DESCRIPTION
**Pull request by cici37**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
This PR is to short circuit remaining validation rules if one exceeds its per-cost limit

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
Part of https://github.com/kubernetes/kubernetes/issues/107573
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go [hunk 1]
```diff
@@ -160,6 +160,10 @@ func (s *Validator) validateExpressions(fldPath *field.Path, sts *schema.Structu
 			continue // todo: wire oldObj parameter
 		}
 		evalResult, evalDetails, err := compiled.Program.Eval(activation)
+		if err != nil {
+			errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("%v evaluating rule: %v", err, ruleErrorString(rule))))
+			continue
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:165]: we still want to handle evalDetails first so we count the cost of the evaluation

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go [hunk 2]
```diff
@@ -176,22 +180,6 @@ func (s *Validator) validateExpressions(fldPath *field.Path, sts *schema.Structu
...
-		if err != nil {
-			// see types.Err for list of well defined error types
-			if strings.HasPrefix(err.Error(), "no such overload") {
-				// Most overload errors are caught by the compiler, which provides details on where exactly in the rule
-				// error was found. Here, an overload error has occurred at runtime no details are provided, so we
-				// append a more descriptive error message. This error can only occur when static type checking has
-				// been bypassed. int-or-string is typed as dynamic and so bypasses compiler type checking.
-				errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("'%v': call arguments did not match a supported operator, function or macro signature for rule: %v", err, ruleErrorString(rule))))
-			} else if strings.HasPrefix(err.Error(), "operation cancelled: actual cost limit exceeded") {
-				errs = append(errs, field.Invalid(fldPath, obj, fmt.Sprintf("'%v': call cost exceeds limit for rule: %v", err, ruleErrorString(rule))))
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:188]: were we expecting to short circuit remaining validation rules if one exceeds its per-cost limit? if so, I do think that was happening, right? We'd decrement remainingCost by the evaluation cost, and include an error here, but not truncate remainingBudget to -1, right?

Comment (cici37) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:188]: Yes. If the cost limit exceed for one expression but the overall cost limit does not exceed, we will keep validating the following rules and return error for this specific failed rule to indicate this rule exceed per expression limit. The CR validation as a whole sure will fail but all the following rules get the chance to be validated and we will return all possible errors together back to user.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go (hunk 1 above)
Comment (cici37) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:165]: With `continue`, the evalDetails will be handled even an error was caught here right?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:165]: no, continue jumps back to the beginning of the for-loop, skipping evalDetails handling

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:188]: hmm, I had been expecting to short circuit further evaluations if we exceeded cost limits in any dimension

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go (hunk 1 above)
Comment (cici37) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go]: So sorry somehow my brain stuck

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go]: > So sorry somehow my brain stuck

welcome to the club... I asked [this question](https://github.com/kubernetes/kubernetes/pull/108419#discussion_r827208071) earlier today, apparently forgetting how if statements work 🙄

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go (hunk 2 above)
Comment (cici37) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:188]: I guess the initial idea of validation is to get all possible check failures so user could fix them at once. We could definitely short circuit further evaluations for cost exceed per expression as well. Not sure which way is better from users' perspective.

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:188]: if it were free, definitely more complete errors are better. if we're bumping up against the protections we put in place for cost limits, I think we err on the side of safety and stop evaluating

Comment (cici37) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go:188]: Make sense. I have updated to short circuit further evaluation when any cost limit exceed. Would you mind to take a look? Thank you!

**Pull request merged**